Dark Bremsstrahlung backport to 10_6_X

revering opened the pull request:
Backport of implementation of Dark Bremsstrahlung physics process from #38329 to CMSSW_10_6_X for run II MC generation. 

CustomPhysics.cc in 10_6_30_patch1 is different from original CMSSW_12_5_X base, but git cms-rebase-topic successfully reapplied the necessary additions for the Dark Brem sim without including other changes. Needed to resolve a merge conflict in CustomPhysics/BuildFile.xml caused by differences in the spacing between \<use name> and the ordering of includes between 10_6_X and 12_5_X. Used the 12_5_X spacing and order, should have no effect (as far as I'm aware).

Generated a small test sample with Dark Brem enabled, and the process biases the cross section and produces Dark photons as expected.

Commit:
Implemented optional Dark Bremsstrahlung process into custom physics

Removed unneccesary include of G4GenericBiasingProcess

perrotta committed:
Update APrimePhysics.h

makortel committed:
Add missing dependencies

cmsbuild commented:
A new Pull Request was created by  @revering  for CMSSW_10_6_X.

It involves the following packages:

- SimG4Core/CustomPhysics (**simulation**)


@cmsbuild, @civanch, @mdhildreth can you please review it and eventually sign? Thanks.
@makortel, @bsunanda, @rovere, @fabiocos, @slomeo this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

civanch commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ccad8b/26540/summary.html
**COMMIT**: ef78fefe0eb3d9d7e3c0da92f7b00dd261172059
**CMSSW**: CMSSW_10_6_X_2022-07-24-0000/slc7_amd64_gcc700
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/38901/26540/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 3215686
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3215350
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 143 log files, 29 edm output root files, 35 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_6_X_2022-07-24-0000+ccad8b/51877/triggerResults): no differences found

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_10_6_X IBs (tests are also fine) and once validation in the development release cycle CMSSW_12_5_X is complete. This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

civanch commented:
@revering , please add in the beginning of the title of this PR: [10_6_X]

civanch commented on SimG4Core/CustomPhysics/src/G4muDarkBremsstrahlungModel.cc:
```diff
@@ -0,0 +1,454 @@
+  for (G4int i = 0; i < nElements; i++) {
+    cross += theAtomNumDensityVector[i] *
+             ComputeCrossSectionPerAtom(
+                 particle, kineticEnergy, (*theElementVector)[i]->GetZ(), (*theElementVector)[i]->GetA(), cut);
```
@revering , better now than never. (*theElementVector)[i]->GetA() returns molar mass of the atom (see https://geant4.kek.jp/lxr/source/materials/include/G4Element.hh). You need to change to
(*theElementVector)[i]->GetN(). The same patch should be added in 12_4 and 12_5.

civanch commented on SimG4Core/CustomPhysics/src/G4muDarkBremsstrahlungModel.cc:
```diff
@@ -0,0 +1,454 @@
+{
+  //Deactivate the process after one dark brem. Needs to be reactivated in the end of event action. If this is in the stepping action instead, more than one brem can occur within each step.
+  G4bool state = false;
+  G4String pname = "muDBrem";
```
@revering , it is a minor thing but better use G4 mass (avoiding problems in corner cases):

double muMass = dp->GetDefinition()->GetPDGMass(); // this is exact muon mass in internal units (MeV)

civanch commented:
@revering , I put two comments to the model. Better to implement this just now and forward port in new PRs to 12_4 and 12_5.

qliphy commented:
backport of #38329

Commit:
Removed obsolete PrintInfo() function from G4muDarkBremsstrahlung

qliphy commented:
@revering   Please note there is a comment raised in master PR: https://github.com/cms-sw/cmssw/pull/38329#issuecomment-1202841384

Commit:
Change G4muDarkBremsstrahlungModel to get muon mass from geant particle definition instead of hardcoded value
and use GetN instead of GetA to get target nucleon number.

Implement code-format fix for G4muDarkBremsstrahlungModel

Change G4muDarkBremsstrahlungModel to use Geant4 muon mass for
muon_mass_mev as well as GeV muon mass.

qliphy commented:
#38949  needs to be backported here also.  @revering

Commit:
Merged DBremSim_10_6_X from repository revering with cms-merge-topic

cmsbuild commented:
Pull request #38901 was updated. @cmsbuild, @civanch, @mdhildreth can you please check and sign again.

qliphy commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ccad8b/26633/summary.html
**COMMIT**: b97a5e72c4068cb657c1dac02be9b0a490a5b06f
**CMSSW**: CMSSW_10_6_X_2022-07-31-0000/slc7_amd64_gcc700
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/38901/26633/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 3215686
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3215350
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 143 log files, 29 edm output root files, 35 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_6_X_2022-07-31-0000+ccad8b/51967/triggerResults): no differences found

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_10_6_X IBs (tests are also fine) and once validation in the development release cycle CMSSW_12_5_X is complete. This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1
